Jetty 12 : Move `FileID` from `jetty-deploy` to `jetty-util`

joakime opened the pull request:
+ Move static methods from other places to `FileID` or `URIUtil`
  - `MultiReleaseJarFile`
  - `Resource`
  - `MetaInfConfiguration`
+ Improve testing of `URIUtil` and `FileID`

joakime committed:
Move FileID from deploy to util

+ Move static methods from other
  places to FileID or URIUtil
  - MultiReleaseJarFile
  - Resource
  - MetaInfConfiguration
+ Improve testing of URIUtil and FileID

joakime committed:
Updating FileID.isTld(Path)

lorban commented on jetty-core/jetty-deploy/src/test/java/org/eclipse/jetty/deploy/MockAppProvider.java:
```diff
@@ -60,9 +60,9 @@ public ContextHandler createContextHandler(App app) throws Exception
         String name = app.getPath().toString();
         name = name.substring(name.lastIndexOf("-")  + 1);
-        File war = new File(webappsDir, name);
+        Path war = new File(webappsDir, name).toPath();
```
Why don't you change the `File webappsDir` field instead?

joakime replied:
Done.

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:
```diff
@@ -0,0 +1,380 @@
+     * @param directoryName the directory name to look for (case insensitive lookup)
+     * @return true if the directory name exists in path, false if otherwise
+     */
+    public static boolean containsDirectory(Path path, String directoryName)
```
This implementation is broken when `path` is `/`.

joakime replied:
This case is protected by the `path.getNameCount()` which returns 0, causing the for loop statements to not run.
I'll add a testcase for this anyway.

lorban replied:
Wouldn't `containsDirectory(Path.get("/"), "tmp")` return false even if `/tmp` exists?

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:
```diff
@@ -0,0 +1,380 @@
+     */
+    public static String getBasename(Path path)
+    {
+        String basename = path.getFileName().toString();
```
You should test for `path.getFileName()` returning `null`.

joakime replied:
Adding testcase

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:
```diff
@@ -0,0 +1,380 @@
+        if (!Files.isRegularFile(path))
+            return null; // not a file
+
+        return getExtension(path.getFileName().toString());
```
You should test for `path.getFileName()` returning `null`.

joakime replied:
Adding testcase

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:
```diff
@@ -0,0 +1,380 @@
+     * @param path the path to test
+     * @return true if not in {@code META-INF/versions/*} tree
+     */
+    public static boolean skipMetaInfVersions(Path path)
```
Rename to `isNotMetaInfVersions`?

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/FileID.java:
```diff
@@ -0,0 +1,380 @@
+     * @param path the path to test
+     * @return true if not a {@code module-info.class} file
+     */
+    public static boolean skipModuleInfoClass(Path path)
```
rename to `isNotModuleInfoClass`?

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/URIUtil.java:
```diff
@@ -1647,25 +1655,7 @@ public static String addQueries(String query1, String query2)
-    }
-
-    public static URI fixBadJavaIoFileUrl(URI uri)
+    public static URI correctBadFileURI(URI uri)
```
Rename to `correctFileURI`?

lorban commented on jetty-core/jetty-util/src/main/java/org/eclipse/jetty/util/resource/Resource.java:
```diff
@@ -852,6 +719,7 @@ public Resource resolve(String subUriPath) throws IOException
     }
 
+    // TODO: move to URIUtil
     private static URI uriResolve(URI uri, String subUriPath) throws IOException
```
Move to `URIUtil` maybe? :-)

joakime replied:
I'm going to get rid of this, as it doesn't make much sense anymore.

lorban commented on jetty-ee10/jetty-ee10-annotations/src/main/java/org/eclipse/jetty/ee10/annotations/AnnotationParser.java:
```diff
@@ -917,7 +917,7 @@ public void resetParsedClasses()
      * @param name the class file name
      * @return whether the class file name is valid
-     * TODO: does multiple things, we should be able to move this FileID in future PR
+     * TODO: does multiple things, we should be able to eliminate this in future PR by consolidating the ZipFS and Directory scanning behaviors into a single Files.walk() stream
```
Do you already have the shape of that future PR in your head?

joakime replied:
Yes.

I want to consolidate the WAR/JAR file scanning/walking with the Directory scanning/walking.
AFAICT, the only difference between the two is that the Directory walking technique also does a "is hidden file" check, which we have a static predicate for in `FileID`, and that same check in WAR walk is harmless.   That means the AnnotationParser gets even simpler.

lorban reviewed (changes requested):
There are a couple small bugs with `FileID` helpers' handling of `/` when `Path.getFilename()` is called, otherwise this looks good.

joakime committed:
Updates from review

lorban reviewed (changes requested):
LGTM, besides one last tiny bug.

joakime committed:
Changing contract of Resource.resolve(String)

No longer throws IOException

joakime committed:
Review of Resource throwables

joakime committed:
Change containsDirectory to hasNamedPathSegment

joakime committed:
Removing unused throwables.

joakime committed:
Fixing missing XML entities

+ Adding missing core XML entities
  to jetty-xml project
  * datatypes.dtd
  * xml.xsd
  * XMLSchema.dtd
+ registering core XML entities
  in XmlParser itself
+ adding new tests
+ Descriptor now catches SAXException
+ WebDescriptor doesn't worry about
  core XML entities

joakime committed:
Restoring jetty-logging.properties

joakime commented on jetty-ee10/jetty-ee10-tests/jetty-ee10-test-quickstart/src/test/resources/jetty-logging.properties:
```diff
@@ -1,3 +1,5 @@
 # Jetty Logging using jetty-slf4j-impl
 org.eclipse.jetty.LEVEL=INFO
+org.eclipse.jetty.xml.LEVEL=DEBUG
+org.eclipse.jetty.quickstart.LEVEL=DEBUG
```
This file was restored/rollback'd in later commit fdbeb2af7f407975b1143df08d2a5e0fbbb29c93

joakime committed:
Make backport of WebDescriptor to EE8 more reliable

joakime committed:
Correcting ee9 web entity url names

joakime committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-move-fileid-to-util

joakime committed:
Rework ee detection code in WebDescriptor

+ Now only loads ee9 specific XML entities if running in an ee9 package.
+ Better calculation of ee level.

joakime committed:
Basing WebDescriptor XML Entity lookup on SERVLET_MAJOR_VERSION instead

joakime commented:
Merged into `jetty-12.0.x`
